### DISCUSSION TITLE
telegram can't handle requests from multiple users and cannot store user data

### DISCUSSION THREAD

**Discussion by simabmagwa** [telegraf/telegraf#1639] [Q&A]
Hello, I've created a telegram bot using telegraf+mysql+node.js . My telegram bot however cannot handle multiple user requests, it mixes them up and after a while it actually loses the data stored (I store data from DB in dataStore=[], it works for a while but after a few requests  I receive a 'dataStore is not defined' error. from the articles I read, it seems my bot is in need of "Sessions and state management" and I honestly don't know if that's the solution and if it is how to implement it. Please help, your assistance and response would be highly appreciated.

**Comment by vedmalex** [comment]
It seems like you’re done something wrong… 
provide code sample to proof this sentence :)

**Reply by simabmagwa** [reply]
I've changed the code and I'm now making use of MySQL powered session middleware for telegraf (telegraf-mysql-session) and it has made my life so much easier ,now I store some of the variables that I used to store in an array using ctx.session. It's so much easier. Thank for your response.

**Reply by Giann8** [reply]
how di you fixed it? I have a similar problem with my gambit, if two or more users start the bot username in are mixed up (it takes the last username and if someone try to do a mission the bot sends message to the wrong person and stores the data in the inventory of that wrong person).
leave you my code of on starting bot:

```
bot.start(async (ctx) => {
    var User = new Character
    User.usern = ctx.message.from.username

    let GetId = () => { return ctx.chat.id }

    if (User.usern == null) {
        return ctx.sendMessage('A quanto pare non hai impostato alcuno username, perfavore vai nelle impostazioni di telegram e crea un nuovo username poi torna e utilizza /start')
    }

    let Player = await dbo.findOne({ usern: User.usern })
    if (Player == null) {
        dbo.updateOne(
            { usern: User.usern },
            { $set: { usern: User.usern, name: ctx.message.from.first_name, level: User.level, exp: User.exp, backpack: User.backpack } },
            { upsert: true }
        )
        return ctx.sendMessage('Hey benvenuto, a quanto pare sei nuovo restarta il bot prima di continuare con /start')
    }
    User = Player

    InGameUsers.set(GetId, User)

    ctx.scene.enter('Università_id')

```

**Reply by spoderx555** [reply]
the problem is with async the function try this code 
```
bot.start((ctx) => {
  (async () => {
    var User = new Character
    User.usern = ctx.message.from.username

    let GetId = () => { return ctx.chat.id }

    if (User.usern == null) {
        return ctx.sendMessage('A quanto pare non hai impostato alcuno username, perfavore vai nelle impostazioni di telegram e crea un nuovo username poi torna e utilizza /start')
    }

    let Player = await dbo.findOne({ usern: User.usern })
    if (Player == null) {
        dbo.updateOne(
            { usern: User.usern },
            { $set: { usern: User.usern, name: ctx.message.from.first_name, level: User.level, exp: User.exp, backpack: User.backpack } },
            { upsert: true }
        )
        return ctx.sendMessage('Hey benvenuto, a quanto pare sei nuovo restarta il bot prima di continuare con /start')
    }
    User = Player

    InGameUsers.set(GetId, User)

    ctx.scene.enter('Università_id')
  })()
```
